### ISSUE TITLE
List curation option - "Limited"/"Group"

### ISSUE THREAD

**Issue by hughrun** [bookwyrm-social/bookwyrm#1318]
**Is your feature request related to a problem? Please describe.**

_Scenario:_

Alex wants to create a list that both Alex and Elham can add books to directly, but nobody else. Currently the only option available is `curated`, but this allows _anyone_ to suggest books and does not allow Elham to add them directly.

**Describe the solution you'd like**
I would like a fourth option called "Limited" or "Group" or something similar, which allows the list owner to add one or more other users as allowed to add books, and prevents anyone else. Basically it's like `public` for the selected users and like `Closed` for everyone else.

**Describe alternatives you've considered**
A possible alternative would be an option to allow anyone a user is following or any followers to add items to a list, however that makes assumptions about the relationship between followers and followees.

**Additional context**

**Comment by hughrun** [comment]
I'd like to work on this. I'm new to Django but I think I can see how this would be done.

**Comment by mouse-reeve** [comment]
Go for it! Let me know if you have any questions

**Comment by mouse-reeve** [comment]
Wow, thank you for all this thought and detail! This rules.

The first thing that comes to mind for this is whether/how to support cross-instance teams. For a team to include a user on a remote instance, there would need to be an activitypub representation of the list members (maybe an OrderedCollection associated with a field on the list object?). In the bigger picture, this would be good, but for now it would be a lot simpler keep the scope of this work just to teams within an instance and not have to handle federation. 

As to questions:
> Q1: is there a better way to design this dropdown, or a better place to put it on the user profile page?

I do think this makes sense to have on the profile page when the viewer has team-managed lists, yeah. I'm not sure where the perfect right place is, but I think this is basically solid

> Q2: Currently after adding a user to a list, the active user is redirected to the list - should they be returned to the user page they were on?

I think redirecting to the list is good.

> Q3: should this be placed in a different location on the List page? & Q4: should it be placed on a different page?

I like your idea to have a separate page from the list itself that shows the team members and has the interface for removing members

> Q5: should team members be allowed to delete books from the list?

I'm leaning to yes, let any member remove a book from the list, but it's an opinion weakly held :)

> Q6: is "Team" the right word to use for this?

Ah, wording, my number one nemesis. My one concern is that it'd be good to keep the phrasing as consistent as possible. Right now both "list members," "team member," and "list team" are in use. I'm not sure if I like "list team" or "list members" better, but I guess "members" is a little more vague?

**Comment by mouse-reeve** [comment]
Oh and one last though, maybe having a panel to search for users to add to the team on the list team page would work in addition to the link on the profile page? Similar to how you can search for a book on the list page.

**Comment by hughrun** [comment]
> I like your idea to have a separate page from the list itself that shows the team members and has the interface for removing members

I don't think this is what I meant to suggest but I agree it's the best approach. If there's a separate page for team management, however, we may as well go all the way and make the team independent of the list. That potentially means:

- a team has its own page, which is where logged-in team "owners" can add and remove members, and users can see who is in the team (note: not sure how user privacy settings would work here)
- a team can provide access to one or more team lists
- teams are associated with team lists
- users are associated with teams rather than with team lists directly (i.e. you are a member of a team, not a member of a team list)
- team owners could choose whether or not team members can delete books from lists (potential future feature)
- team owners could choose whether or not team members can create team lists assigned to the team (potential future feature)
- other future "team" features could be possible (e.g. something useful to book groups, or collective ownership of the team somehow)

This might also help with the naming problem.

Then:
- a **team** can be created from the user page in a similar process to creating a list
- a team has a **team name**, **team lists** that members can contribute to, and Users as **team members** (including the owner, which would make notifications and templates logic cleaner)
- owners go to the team page to add and remove other users to/from the **team** (with user search as @mouse-reeve suggests)
- users must be a member of one or more teams before they can create **team lists** in the list creation screen
- when creating a team list, users would select the team from a dropdown when creating the list
- a button somewhere on the list page would link to the associated team page, if it is a team list
- possibly change the (new) `Team Lists` tab on user profiles to `Teams` instead so we list the user's teams rather than the `team lists` they have access to

Hopefully that all is clear - but is it _sensible_? This is a fair way beyond my original proposal but could provide a more flexible platform for extending functionality in future.

**Comment by hughrun** [comment]
Ok I nearly have a PR ready for this. Just sorting out invitations and notifications - should be ready in a couple of days.

I probably will also need some help making `privacy_filter` work with group memberships.

**Comment by mouse-reeve** [comment]
Wonderful! And yeah that function is kinda gnarly :)
